### DISCUSSION TITLE
using normal maps

### DISCUSSION THREAD

**Discussion by jasonswearingen** [pink-arcana/godot-distance-field-outlines#7] [General]
first of all, thank you for this resource.   I have not done any shader/gfx dev before but this will be my starting point once I get my gameplay prototype done.

I was wondering how difficult it would be to add screen-space normalmaps / depth buffer into the computation?  so you can get things like interrior edges, and catching the boundary between different geometries that have the same color.

Or is doing that basically just doing things that other screen-space edge-enhancement shaders do?

**Comment by pink-arcana** [comment]
Hi! Yeah, so, this project is definitely set up for you to be able to add your own extraction algirithm using other methods, including normals and depth. You would do it exactly same way that other screen-space edge shaders do. There's info in the [Shaders](https://github.com/pink-arcana/godot-distance-field-outlines?tab=readme-ov-file#shaders) section of the README. 

For the DFOutlineCE, you'll need to use [this equation to get normals data from the normal roughness buffer](https://github.com/pink-arcana/godot-distance-field-outlines/discussions/1#discussioncomment-10016608), and for DFOutlineNode you'd need to use [the method here](https://github.com/pink-arcana/godot-distance-field-outlines/blob/main/project/df_outline_node/README.md#adding-additional-buffers). 

You're not the first person who's expressed interest in them, and, because getting the buffers isn't as straightforward as I'd hoped, I am thinking of adding these buffers into the extraction shaders by  default in an upcoming version. 

However, I'm not sure that normals-based outlines will be a viable choice for most games. They tend to have artifacts that flicker frame-to-frame where the angles in the geometry are close to the threshold for outlines. And, in a shader like this that creates very prominent wide outlines, the flickering looks exponentially worse.

For this reason, I will be adding a demo scene that uses Object ID and UV ID buffers. This is what I have in my current game and is the only way I've found to have clean, reliable outlines that don't flicker. The UV ID buffer in particular requires significant manual work to setup (using vertex colors or UV2), but it gives you reliable detailed outlines over internal geometry.

**Comment by jasonswearingen** [comment]
Thank you for the reply.  I didn't know object/uv buffers were a thing (all my knowledge is book knowledge, from like DX9), but your explanation makes it seem totally the way to do it well.  This is so exciting!  I need to hurry my ass up :)

**Comment by pink-arcana** [comment]
I went ahead and added normals and depth buffers to the CompositorEffect's extraction shader. This means you should be able to port any quadmesh screen-space outline shaders to it without having to edit the CompositorEffect. If you end up trying it out, let me know how it works!

Instructions here: https://github.com/pink-arcana/godot-distance-field-outlines/tree/main/project/df_outline_ce#customizing-outline-extraction

**Comment by jasonswearingen** [comment]
a bit of a ramble, just me thinking through things, feel free to ignore. 

I've started experimenting with environment/worldgen finally, and so am thinking about how to apply your DF techniques more....

I am thinking of procedural placement of "blocks".  not really in a voxel context but more of just kitbash.

I think for same color shapes, probably your current technique with some depth+normal based enhancement would be better than object/uv id's.   That way clusters of "the same" block placed in close proximity would be treated as the same object from an outline perspective.  

a disclaimer though, I don't know how OBjectID's or UV ID's would work, for example if repetitive placements of the same block could have the same object/uv id, then that could be fine also.   I keep thinking that depth+normal does still have a place, for detecting interior edges at minimum.    but maybe more the domain of a different technque, not in a full screen postprocess.

**Comment by pink-arcana** [comment]
Let me know what ends up working out! Depth/normals are obviously a popular choice for outlines. And they can look really good. But I've never seen an implementation that was able to avoid flickering, and my eyes are pretty sensitive to flickering artifacts, so that's where I'm coming from.

The object/surface IDs do come at a huge cost of complexity and extra work preparing meshes, so they're definitely not right for everyone. ([My WIP extension of this project](https://www.reddit.com/r/godot/comments/1ebvmvt/coming_soon_advanced_distance_field_outlines_mit/) to use object/surfaces is far more complex than the distance field CompositorEffect here.) But the input data is very precise, so the depth sorting and fade is much more accurate. And having object ID SDFs give you a lot more control than a normal screen-space shader would have. But, again, that amount of control isn't needed for all (probably most) projects.

As far as the actual IDs, there are different ways you could implement them. All I know about it is from a couple game devlogs, so I'm mostly winging it. My setup automatically assigns unique IDs as instance uniforms to each mesh (but multiple meshes can share an ID if they're supposed to be outlined as a single object). And, for adjacent surfaces, it only draws a line if they are on the same object, so it's not a problem to have the same meshes side-by-side.
